### PR TITLE
Add routing-java-ortools project

### PR DESCRIPTION
**Pull request by larsbeck**
# Description

Adds an ORTools based routing app in Java. It mirrors the functionality of the `routing-ortools` app in Python.

### CODE REVIEW

File: routing-java-ortools/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Java OR-Tools Maven Example
+
+Sample project explaining how to use Google OR-Tools in a maven project that can
+be executed on the Nextmv platform.
+
+## Usage
+
+Java templates in Nextmv require a `main.jar` as an entry point. Running the
```
Comment (merschformann) [inline routing-java-ortools/README.md]:
```suggestion
Java apps in Nextmv require a `main.jar` as an entry point. Running the
```

Seems nicer.

File: routing-java-ortools/VERSION [hunk 2]
```diff
@@ -0,0 +1 @@
+v1.2.1
```
Comment (merschformann) [inline routing-java-ortools/VERSION]: Why are we starting this app at version `v1.2.1`? :thinking:

File: routing-java-ortools/src/main/java/com/nextmv/example/Main.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+import com.google.ortools.constraintsolver.main;
+import com.google.protobuf.Duration;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public final class Main {
+  public static void main(String[] args) {
+    // record start time of the program
```
Comment (merschformann) [inline routing-java-ortools/src/main/java/com/nextmv/example/Main.java]:
```suggestion
    // Record start time of the program.
```

Quick comment style alignment.

File: routing-java-ortools/src/main/java/com/nextmv/example/Main.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    long startTime = System.currentTimeMillis();
+    // Parse arguments. Exit on error.
+    Options options = Options.fromArguments(args);
+
+    // Load input. Exit on error.
+    Input input = Input.fromString(options.getInputPath());
+
+    Loader.loadNativeLibraries();
+
+    // Create Routing Index Manager
```
Comment (merschformann) [inline routing-java-ortools/src/main/java/com/nextmv/example/Main.java]:
```suggestion
    // Create Routing Index Manager.
```

File: routing-java-ortools/src/main/java/com/nextmv/example/Main.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+      // Convert from routing variable Index to user NodeIndex.
+      int fromNode = manager.indexToNode(fromIndex);
+      int toNode = manager.indexToNode(toIndex);
+      return input.distanceMatrix[fromNode][toNode];
+    });
+
+    // Define cost of each arc.
+    routing.setArcCostEvaluatorOfAllVehicles(transitCallbackIndex);
+
+    // set the duration of the search
```
Comment (merschformann) [inline routing-java-ortools/src/main/java/com/nextmv/example/Main.java]:
```suggestion
    // Set the duration of the search.
```

File: routing-java-ortools/src/main/java/com/nextmv/example/Main.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+    // Setting first solution heuristic.
+    RoutingSearchParameters searchParameters = main.defaultRoutingSearchParameters()
+        .toBuilder()
+        .setFirstSolutionStrategy(FirstSolutionStrategy.Value.PATH_CHEAPEST_ARC)
+        .setTimeLimit(duration)
+        .build();
+
+    // Solve the problem.
+    // record solve start time
```
Comment (merschformann) [inline routing-java-ortools/src/main/java/com/nextmv/example/Main.java]:
```suggestion
    // Record solve start time
```

File: routing-java-ortools/src/main/java/com/nextmv/example/Main.java [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    // convert duration to seconds
+    duration = duration / 1000.0;
+
+    // compute total duration
+    endTime = System.currentTimeMillis();
+    double runDuration = endTime - startTime;
+    // convert duration to seconds
+    runDuration = runDuration / 1000.0;
+
+    // create output
```
Comment (merschformann) [inline routing-java-ortools/src/main/java/com/nextmv/example/Main.java]:
```suggestion
    // Compute solve duration.
    long endTime = System.currentTimeMillis();
    double duration = endTime - solveStartTime;
    // Convert duration to seconds.
    duration = duration / 1000.0;

    // Compute total duration.
    endTime = System.currentTimeMillis();
    double runDuration = endTime - startTime;
    // Convert duration to seconds.
    runDuration = runDuration / 1000.0;

    // Create output.
```

What have I started? :sweat_smile:

File: routing-java-ortools/src/main/java/com/nextmv/example/Output.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      
+    // find the vehicle with the maximum route distance
+    this.statistics.result.custom.maxRouteDistance = (int) vehicles.stream()
+        .mapToDouble(v -> v.getDistance()-2).max().orElse(0);
+
+    // find the vehicle with the maximum number of stops
+    this.statistics.result.custom.maxStopsInVehicle = (int) vehicles.stream()
+        .mapToDouble(v -> v.getStops().size()-2).max().orElse(0);
+
+    // find the vehicle with the minimum number of stops
```
Comment (merschformann) [inline routing-java-ortools/src/main/java/com/nextmv/example/Output.java]:
```suggestion
    // Fill custom section.
    this.statistics.result.custom = new StatisticsResultCustom();

    // A vehicle is activated if it has at least three stops. The first and last
    // stop are the depot, so the vehicle has at least one customer stop.
    this.statistics.result.custom.activatedVehicles = (int) vehicles.stream()
        .filter(v -> v.getStops().size() > 3).count();
      
    // Find the vehicle with the maximum route distance.
    this.statistics.result.custom.maxRouteDistance = (int) vehicles.stream()
        .mapToDouble(v -> v.getDistance()-2).max().orElse(0);

    // Find the vehicle with the maximum number of stops.
    this.statistics.result.custom.maxStopsInVehicle = (int) vehicles.stream()
        .mapToDouble(v -> v.getStops().size()-2).max().orElse(0);

    // Find the vehicle with the minimum number of stops.
```

Over-committed.

**Review by merschformann — APPROVED** [review]
There are some further steps to do if we want to make it a marketplace app (see [here](https://github.com/nextmv-io/community-apps/blob/develop/.nextmv/workflow-configuration.yml)). 
However, potentially not an immediate action.

File: routing-java-ortools/VERSION (hunk 2 above)
Comment (larsbeck) [reply routing-java-ortools/VERSION]: Good catch. Set it to 1.0.0

**Pull request merged**